File: voice/tts/tts_cpp/tts.cpp
// tts.cpp - Matcha TTS 中文语音合成 (sherpa-onnx C API)
//
// 用法:
//   ./tts "你好，欢迎使用语音合成"
//   ./tts -o hello.wav -s 1.2 "100人参加活动"
//   ./tts --model-dir=./models --num-threads=4 "你的文本"

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <getopt.h>
#include "sherpa-onnx/c-api/c-api.h"

struct Config {
    std::string model_dir = "./matcha-icefall-zh-baker";
    std::string vocoder;
    std::string output = "tts_output.wav";
    std::string text;
    float speed = 1.0f;
    int num_threads = 2;
    bool debug = false;
};

void usage(const char *prog) {
    fprintf(stderr,
        "Matcha TTS 中文语音合成 (sherpa-onnx C API)\n\n"
        "用法: %s [选项] \"要合成的文本\"\n\n"
        "选项:\n"
        "  -o, --output FILE     输出 WAV (默认: tts_output.wav)\n"
        "  -s, --speed FLOAT     语速 >1快 <1慢 (默认: 1.0)\n"
        "  -m, --model-dir DIR   模型目录 (默认: ./matcha-icefall-zh-baker)\n"
        "  -v, --vocoder FILE    声码器路径 (默认: model-dir/vocos-22khz-univ.onnx)\n"
        "  -t, --num-threads N   线程数 (默认: 2)\n"
        "  -d, --debug           显示调试信息\n"
        "  -h, --help            显示帮助\n\n"
        "示例:\n"
        "  %s \"你好世界\"\n"
        "  %s -o hello.wav -s 1.2 \"100人参加活动\"\n",
        prog, prog, prog);
}

int parse_args(int argc, char *argv[], Config &cfg) {
    static struct option opts[] = {
        {"output",      required_argument, 0, 'o'},
        {"speed",       required_argument, 0, 's'},
        {"model-dir",   required_argument, 0, 'm'},
        {"vocoder",     required_argument, 0, 'v'},
        {"num-threads", required_argument, 0, 't'},
        {"debug",       no_argument,       0, 'd'},
        {"help",        no_argument,       0, 'h'},
        {0,0,0,0}
    };
    int c;
    while ((c = getopt_long(argc, argv, "o:s:m:v:t:dh", opts, NULL)) != -1) {
        switch (c) {
            case 'o': cfg.output = optarg; break;
            case 's': cfg.speed = atof(optarg); break;
            case 'm': cfg.model_dir = optarg; break;
            case 'v': cfg.vocoder = optarg; break;
            case 't': cfg.num_threads = atoi(optarg); break;
            case 'd': cfg.debug = true; break;
            case 'h': usage(argv[0]); exit(0);
            default:  usage(argv[0]); return -1;
        }
    }
    if (optind < argc) cfg.text = argv[optind];
    if (cfg.text.empty()) {
        fprintf(stderr, "错误: 未提供文本\n");
        usage(argv[0]);
        return -1;
    }
    if (cfg.vocoder.empty())
        cfg.vocoder = cfg.model_dir + "/vocos-22khz-univ.onnx";
    return 0;
}

static int progress_cb(const float*, int32_t, float p, void*) {
    fprintf(stderr, "\r进度: %.1f%%", p * 100);
    fflush(stderr);
    return 1;
}

int main(int argc, char *argv[]) {
    Config cfg;
    if (parse_args(argc, argv, cfg) != 0) return 1;

    auto acoustic = cfg.model_dir + "/model-steps-3.onnx";
    auto lexicon  = cfg.model_dir + "/lexicon.txt";
    auto tokens   = cfg.model_dir + "/tokens.txt";
    auto dict_dir = cfg.model_dir + "/dict";
    auto rules    = cfg.model_dir + "/phone.fst," +
                    cfg.model_dir + "/date.fst," +
                    cfg.model_dir + "/number.fst";

    SherpaOnnxOfflineTtsConfig c;
    memset(&c, 0, sizeof(c));
    c.model.matcha.acoustic_model = acoustic.c_str();
    c.model.matcha.vocoder        = cfg.vocoder.c_str();
    c.model.matcha.lexicon        = lexicon.c_str();
    c.model.matcha.tokens         = tokens.c_str();
    c.model.matcha.dict_dir       = dict_dir.c_str();
    c.model.num_threads           = cfg.num_threads;
    c.model.debug                 = cfg.debug ? 1 : 0;
    c.rule_fsts                   = rules.c_str();

    auto *tts = SherpaOnnxCreateOfflineTts(&c);
    if (!tts) { fprintf(stderr, "错误: TTS 创建失败\n"); return 1; }

    SherpaOnnxGenerationConfig gc = {0};
    gc.sid = 0;
    gc.speed = cfg.speed;
    gc.silence_scale = 0.2f;

    fprintf(stderr, "文本: %s\n", cfg.text.c_str());
    auto *audio = SherpaOnnxOfflineTtsGenerateWithConfig(
        tts, cfg.text.c_str(), &gc, progress_cb, NULL);
    fprintf(stderr, "\n");

    if (!audio || audio->n == 0) {
        fprintf(stderr, "错误: 生成失败\n");
        SherpaOnnxDestroyOfflineTts(tts);
        return 1;
    }

    SherpaOnnxWriteWave(audio->samples, audio->n, audio->sample_rate,
                        cfg.output.c_str());
    fprintf(stderr, "已保存: %s (%.2fs, %dHz)\n",
            cfg.output.c_str(),
            (float)audio->n / audio->sample_rate,
            audio->sample_rate);

    SherpaOnnxDestroyOfflineTtsGeneratedAudio(audio);
    SherpaOnnxDestroyOfflineTts(tts);
    return 0;
}

File: voice/tts/tts_py/tts.py
#!/usr/bin/env python3
"""
Matcha TTS 中文语音合成脚本
基于 sherpa-onnx，支持自定义文本、语速、输出文件

用法:
  python3 tts.py "你好，欢迎使用语音合成系统"
  python3 tts.py --output=hello.wav --speed=1.2 "你好世界"
  python3 tts.py --num-threads=4 "很长的文本..."
"""

import argparse
import time
import os
import sys

try:
    import sherpa_onnx
    import soundfile as sf
except ImportError:
    print("错误: 请先安装依赖")
    print("  pip install sherpa-onnx soundfile")
    sys.exit(1)


def get_args():
    parser = argparse.ArgumentParser(
        description="Matcha TTS 中文语音合成",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    parser.add_argument(
        "text",
        type=str,
        help="要合成的文本",
    )
    parser.add_argument(
        "--output", "-o",
        type=str,
        default="tts_output.wav",
        help="输出 WAV 文件路径",
    )
    parser.add_argument(
        "--speed",
        type=float,
        default=1.0,
        help="语速，越大越快，越小越慢",
    )
    parser.add_argument(
        "--num-threads",
        type=int,
        default=2,
        help="线程数",
    )
    parser.add_argument(
        "--model-dir",
        type=str,
        default="./matcha-icefall-zh-baker",
        help="模型目录路径",
    )
    parser.add_argument(
        "--vocoder",
        type=str,
        default="./matcha-icefall-zh-baker/vocos-22khz-univ.onnx",
        help="声码器模型路径",
    )
    parser.add_argument(
        "--debug",
        action="store_true",
        help="显示调试信息",
    )
    return parser.parse_args()


def main():
    args = get_args()

    model_dir = args.model_dir
    vocoder = args.vocoder

    # 检查模型文件是否存在
    required_files = [
        os.path.join(model_dir, "model-steps-3.onnx"),
        vocoder,
        os.path.join(model_dir, "lexicon.txt"),
        os.path.join(model_dir, "tokens.txt"),
        os.path.join(model_dir, "phone.fst"),
        os.path.join(model_dir, "date.fst"),
        os.path.join(model_dir, "number.fst"),
    ]
    for f in required_files:
        if not os.path.exists(f):
            print(f"错误: 模型文件不存在: {f}")
            sys.exit(1)

    dict_dir = os.path.join(model_dir, "dict")
    if not os.path.isdir(dict_dir):
        print(f"错误: 字典目录不存在: {dict_dir}")
        sys.exit(1)

    # 构建 TTS 配置
    tts_config = sherpa_onnx.OfflineTtsConfig(
        model=sherpa_onnx.OfflineTtsModelConfig(
            matcha=sherpa_onnx.OfflineTtsMatchaModelConfig(
                acoustic_model=os.path.join(model_dir, "model-steps-3.onnx"),
                vocoder=vocoder,
                lexicon=os.path.join(model_dir, "lexicon.txt"),
                tokens=os.path.join(model_dir, "tokens.txt"),
            ),
            num_threads=args.num_threads,
            debug=args.debug,
            provider="cpu",
        ),
        rule_fsts=",".join([
            os.path.join(model_dir, "phone.fst"),
            os.path.join(model_dir, "date.fst"),
            os.path.join(model_dir, "number.fst"),
        ]),
        max_num_sentences=1,
    )

    if not tts_config.validate():
        print("错误: TTS 配置无效，请检查模型路径")
        sys.exit(1)

    # 创建 TTS 实例
    tts = sherpa_onnx.OfflineTts(tts_config)

    # 生成语音
    print(f"合成文本: {args.text}")
    print(f"语速: {args.speed}")

    start = time.time()
    gen_config = sherpa_onnx.GenerationConfig()
    gen_config.sid = 0
    gen_config.speed = args.speed
    gen_config.silence_scale = 0.2

    audio = tts.generate(args.text, gen_config)
    elapsed = time.time() - start

    if len(audio.samples) == 0:
        print("错误: 语音生成失败")
        sys.exit(1)

    # 保存 WAV
    sf.write(
        args.output,
        audio.samples,
        samplerate=audio.sample_rate,
        subtype="PCM_16",
    )

    duration = len(audio.samples) / audio.sample_rate
    rtf = elapsed / duration if duration > 0 else 0

    print(f"已保存到: {args.output}")
    print(f"音频时长: {duration:.2f}s")
    print(f"耗时: {elapsed:.2f}s")
    print(f"RTF: {rtf:.3f}")


if __name__ == "__main__":
    main()
